### PR TITLE
RMLNQSQL-201 Add reconciliation of set operation queries (UNION/UNION ALL)

### PR DESCRIPTION
**Pull request by a-ctor**

### CODE REVIEW

File: UnitTests/MappingResolution/DefaultMappingResolutionStageTest.cs [hunk 1]
```diff
@@ -409,6 +410,84 @@ public void ResolveMemberAccess ()
       Assert.That (result, Is.SameAs (fakeResolvedExpression));
     }
     
+    [Test]
+    public void ResolveSetOperationReconciliationContext_ResolverReturnsTrue_ReturnsReconciliationContext ()
+    {
+      var projectionExpressions = new Expression[] { Expression.Constant (0), Expression.Constant (1) };
+      var fakeReconciliationContext = Mock.Of<ISetOperationReconciliationContext> ();
```
Comment (MichaelKetting) [inline UnitTests/MappingResolution/DefaultMappingResolutionStageTest.cs]:
```suggestion
      var fakeReconciliationContext = Mock.Of<ISetOperationReconciliationContext>();
```

File: UnitTests/MappingResolution/DefaultMappingResolutionStageTest.cs [hunk 2]
```diff
@@ -409,6 +410,84 @@ public void ResolveMemberAccess ()
...
+
+      _resolverMock.Verify();
+      Assert.That (result, Is.SameAs (fakeReconciliationContext));
+    }
+
+    [Test]
+    public void ResolveSetOperationReconciliationContext_ResolverReturnsFalse_ReturnsNull ()
+    {
+      var projectionExpressions = new Expression[] { Expression.Constant (0), Expression.Constant (1) };
+      ISetOperationReconciliationContext outContext = null;
```
Comment (MichaelKetting) [inline UnitTests/MappingResolution/DefaultMappingResolutionStageTest.cs]: Please make this a not-null value so we can verify that we do actually return NULL instead of the out-result

File: UnitTests/MappingResolution/DefaultMappingResolutionStageTest.cs [hunk 3]
```diff
@@ -409,6 +410,84 @@ public void ResolveMemberAccess ()
...
+          .Returns (() => (SqlColumnExpression[]) columns.Clone())
+          .Verifiable();
+
+      var result = _stage.ApplySetOperationReconciliationContext (entityExpression, reconciliationContextMock.Object);
+
+      reconciliationContextMock.Verify();
+      Assert.That (result.Columns.Count, Is.EqualTo(2));
+      Assert.That (result.Columns[0], Is.SameAs (columns[0]));
+      Assert.That (result.Columns[1], Is.SameAs (columns[1]));
+    }
```
Comment (MichaelKetting) [inline UnitTests/MappingResolution/DefaultMappingResolutionStageTest.cs]: Test design: if we use a collection assert with Is.Equal, we sometimes get better information in case of fails and a shorter test. It's just something about preference for future tests.

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+          {
+            throw new ArgumentException (
+                $"The column type '{type}' does not match up with the type of previous columns '{entry.Column.Type}'.",
+                nameof(type));
+          }
+        }
+      }
+
+      public Column AddEntry (ColumnEntry entry)
+      {
```
Comment (MichaelKetting) [inline Core/MappingResolution/DefaultSetOperationReconciliationContext.cs:111]: arg-null check please

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+      public SqlEntityExpression Entity { get; }
+
+      public SqlColumnExpression Column { get; }
+
+      public ColumnEntry (SqlEntityExpression entity, SqlColumnExpression column)
+      {
+        ArgumentUtility.CheckNotNull (nameof(entity), entity);
+        ArgumentUtility.CheckNotNull (nameof(column), column);
+        if (!entity.Columns.Contains (column))
+          throw new ArgumentException ($"The specified column '{column}' does not belong to the specified entity.", nameof(column));
```
Comment (MichaelKetting) [inline Core/MappingResolution/DefaultSetOperationReconciliationContext.cs]:
```suggestion
        if (!entity.Columns.Contains (column))
          throw new ArgumentException ($"The column '{column}' does not belong to the entity '{entity.Name}'.", nameof(column));
```

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+          if (columnIndexLookup.ContainsKey (columnEntry.Column))
+          {
+            throw new InvalidOperationException (
+                $"The column '{columnEntry.Column}' is used in multiple times in the same reconciliation context.");
+          }
+
+          columnIndexLookup.Add (columnEntry.Column, columnIndex);
+        }
+
+        columnIndex += 1;
```
Comment (MichaelKetting) [inline Core/MappingResolution/DefaultSetOperationReconciliationContext.cs:171]: Stylistic question: I ually see `++` instead of `+= 1` for incrementing. Is there a semantic intention that's communicated?

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.Linq.SqlBackend.SqlStatementModel.Resolved;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.MappingResolution
+{
+  public class DefaultSetOperationReconciliationContext : ISetOperationReconciliationContext
+  {
+    public class Builder
```
Comment (MichaelKetting) [inline Core/MappingResolution/DefaultSetOperationReconciliationContext.cs]: Since the nested types don't contain virtual members, pelase seal.

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.Linq.SqlBackend.SqlStatementModel.Resolved;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.MappingResolution
+{
+  public class DefaultSetOperationReconciliationContext : ISetOperationReconciliationContext
```
Comment (MichaelKetting) [inline Core/MappingResolution/DefaultSetOperationReconciliationContext.cs:33]: pelase add xml docs and a short explanation when and how to derive from this type

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            column.Name,
+            false);
+      }
+
+      foreach (var column in entityExpression.Columns)
+      {
+        if (!ColumnIndexLookup.TryGetValue (column, out var columnIndex))
+        {
+          throw new InvalidOperationException (
+              $"Column '{column}' is not supported in set operation reconciliation context that claims to support the entity.");
```
Comment (MichaelKetting) [inline Core/MappingResolution/DefaultSetOperationReconciliationContext.cs]:
```suggestion
              $"Column '{column}' is not supported in set operation reconciliation context that claims to support the entity '{entty.Name}'.");
```

File: Core/MappingResolution/IMappingResolutionStage2.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+//
+using System.Linq.Expressions;
+using JetBrains.Annotations;
+using Remotion.Linq.SqlBackend.SqlStatementModel.Resolved;
+
+namespace Remotion.Linq.SqlBackend.MappingResolution
+{
+  /// <summary>
+  /// Extends <see cref="IMappingResolutionStage"/> functionality to prevent breaking changes.
+  /// Gets removed with the next major version.
```
Comment (MichaelKetting) [inline Core/MappingResolution/IMappingResolutionStage2.cs]:
```suggestion
  /// Will be merged into <see cref="IMappingResolutionStage"/> with the next major version.
```

File: Core/MappingResolution/IMappingResolver2.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+//
+using System.Linq.Expressions;
+
+namespace Remotion.Linq.SqlBackend.MappingResolution
+{
+  /// <summary>
+  /// Extends <see cref="IMappingResolver"/> functionality to prevent breaking changes.
+  /// Gets removed with the next major version.
```
Comment (MichaelKetting) [inline Core/MappingResolution/IMappingResolver2.cs]:
```suggestion
  /// Will be merged into <see cref="IMappingResolver"/> with the next major version.
```

File: Core/MappingResolution/SetOperationReconciliationVisitor.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  /// <see cref="SetOperationReconciliationVisitor"/> the selection projections in set operations to match a reconciled column view
+  /// defined by <see cref="ISetOperationReconciliationContext"/>.
+  /// </summary>
+  public class SetOperationReconciliationVisitor
+      : RelinqExpressionVisitor,
+          IResolvedSqlExpressionVisitor
+  {
+    private readonly ISetOperationReconciliationContext _reconciliationContext;
+    private readonly IMappingResolutionStage2 _stage;
+
```
Comment (MichaelKetting) [inline Core/MappingResolution/SetOperationReconciliationVisitor.cs]: Move instance fields  to after the static methods please.

File: UnitTests/MappingResolution/SetOperationReconciliationVisitorTest.cs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    [Test]
+    public void ApplyReconciliation_SqlEntity_NotRequiringReconciliation_ReturnsSameExpression ()
+    {
+      var entityExpression = SqlStatementModelObjectMother.CreateSqlEntityDefinitionExpression (typeof (Cook));
+
+      _reconciliationContextMock
+          .Setup (mock => mock.IsReconciliationRequired (entityExpression))
+          .Returns (false)
+          .Verifiable();
+
```
Comment (MichaelKetting) [inline UnitTests/MappingResolution/SetOperationReconciliationVisitorTest.cs:73]: Please add a stub that throws for ApplySetOperationReconciliationContext .
Reasoning: we could write the code so it defaults to the original result in case of NULL and that unfortunate code change would not be caught by the tests.  I do try to not test for stuff that's not happening but some accidental code changes are more likely than other.

File: Core/SqlStatementModel/Resolved/ISqlColumnExpressionVisitor2.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+// 
+
+using System;
+using System.Linq.Expressions;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel.Resolved
+{
+  public interface ISqlColumnExpressionVisitor2 : ISqlColumnExpressionVisitor
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/ISqlColumnExpressionVisitor2.cs:28]: please add same note as the other `2` interfaces

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs [hunk 15]
```diff
@@ -77,6 +77,21 @@ public void GenerateSql_VisitSqlColumnDefinitionExpression ()
       Assert.That (_commandBuilder.GetCommandText(), Is.EqualTo ("[s].[ID]"));
     }
 
+    [Test]
+    public void GenerateSql_VisitSqlConstantColumnExpression ()
```
Comment (MichaelKetting) [inline UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs]: please add a second test for null and a third test for string that shows we need to take case of the quotes.

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Linq.Expressions;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel.Resolved
+{
+  /// <summary>
+  /// Defines a SQL column as NULL expression column with a given name. The column is represented as "NULL AS [name]".
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: please update docs.

File: Core/SqlGeneration/SqlGeneratingExpressionVisitor.cs [hunk 17]
```diff
@@ -106,6 +107,17 @@ public virtual Expression VisitSqlColumnReference (SqlColumnReferenceExpression
       return expression;
     }
 
+    public virtual Expression VisitSqlConstantColumn (SqlConstantColumnExpression expression)
+    {
+      ArgumentUtility.CheckNotNull (nameof(expression), expression);
+
+      _commandBuilder.Append (expression.Value);
```
Comment (MichaelKetting) [inline Core/SqlGeneration/SqlGeneratingExpressionVisitor.cs]:
```suggestion
      Visit (expression.Value);
```

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+          unsafeValue,
+          type,
+          owningTableAlias,
+          columnName,
+          isPrimaryKey);
+    }
+
+    public string Value { get; }
+
+    private SqlConstantColumnExpression (string value, Type type, string owningTableAlias, string columnName, bool isPrimaryKey)
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]:
```suggestion
    private SqlConstantColumnExpression (Expression value, Type type, string owningTableAlias, string columnName, bool isPrimaryKey)
```

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+      return new SqlConstantColumnExpression (
+          unsafeValue,
+          type,
+          owningTableAlias,
+          columnName,
+          isPrimaryKey);
+    }
+
+    public string Value { get; }
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]:
```suggestion
    public Expression Value { get; }
```

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+  /// </summary>
+  public class SqlConstantColumnExpression : SqlColumnExpression
+  {
+    public static SqlConstantColumnExpression CreateNull (Type type, string owningTableAlias, string columnName, bool isPrimaryKey)
+    {
+      ArgumentUtility.CheckNotNull (nameof(type), type);
+      ArgumentUtility.CheckNotNull (nameof(owningTableAlias), owningTableAlias);
+      ArgumentUtility.CheckNotNull (nameof(columnName), columnName);
+
+      return CreateVerbatimUnsafe ("NULL", type, owningTableAlias, columnName, isPrimaryKey);
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]:
```suggestion
      return new ConstantExpression (new ConstantExpression(null), type, owningTableAlias, columnName, isPrimaryKey);
```

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+      ArgumentUtility.CheckNotNull (nameof(owningTableAlias), owningTableAlias);
+      ArgumentUtility.CheckNotNull (nameof(columnName), columnName);
+
+      return new SqlConstantColumnExpression (
+          unsafeValue,
+          type,
+          owningTableAlias,
+          columnName,
+          isPrimaryKey);
+    }
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]:
```suggestion
    public static SqlConstantColumnExpression CreateConstant (
        object value,
        Type type,
        string owningTableAlias,
        string columnName,
        bool isPrimaryKey)
    {
      ArgumentUtility.CheckNotNull (nameof(value), value);
      ArgumentUtility.CheckNotNull (nameof(type), type);
      ArgumentUtility.CheckNotNull (nameof(owningTableAlias), owningTableAlias);
      ArgumentUtility.CheckNotNull (nameof(columnName), columnName);
if (value is string) throw ;
      return new SqlConstantColumnExpression (
          new ConstantExpression(value),
          type,
          owningTableAlias,
          columnName,
          isPrimaryKey);
    }
    
     public static SqlConstantColumnExpression CreateSqlLiteral (
        string value,
        Type type,
        string owningTableAlias,
        string columnName,
        bool isPrimaryKey)
    {
      ArgumentUtility.CheckNotNull (nameof(unsafeValue), unsafeValue);
      ArgumentUtility.CheckNotNull (nameof(type), type);
      ArgumentUtility.CheckNotNull (nameof(owningTableAlias), owningTableAlias);
      ArgumentUtility.CheckNotNull (nameof(columnName), columnName);

      return new SqlConstantColumnExpression (
          new SqlLiteralExpreession(value),
          type,
          owningTableAlias,
          columnName,
          isPrimaryKey);
    }
```

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+using System.Linq.Expressions;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel.Resolved
+{
+  /// <summary>
+  /// Defines a SQL column as NULL expression column with a given name. The column is represented as "NULL AS [name]".
+  /// This is useful for set operations where two different results are padded to ensure the columns match.
+  /// </summary>
+  public class SqlConstantColumnExpression : SqlColumnExpression
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: Between the comments here and in SqlGeneratingExpressionVIsitor I found a way to handle the variaous onstants in a fashion better aligned with he existing sqlbackend-patterns 

Basically, we use the already exisitng constantExpression and SqlLiteralExpression to handle the magic.

File: UnitTests/TestDomain/Server.cs [hunk 23]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Remotion.Linq.SqlBackend.UnitTests.TestDomain
+{
+  public class Server : Cook
```
Comment (MichaelKetting) [inline UnitTests/TestDomain/Server.cs:3]: Thanks. I had no idea that `server` is actually in use as a common alternative to `waiter`

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs (hunk 22 above)
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: With the other changes, should we rename this type to SqlNamedProjectionColumnExpression or something of the sort? I did not google the proper terms. I do think it's possible to inject more magic than just a constant value this way and there is no reason to limit the type after the other changes are done.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
I found a solution for the contant-value column. I think the comments should be suffcient and the update contained and straight forward.

File: Core/MappingResolution/DefaultSetOperationReconciliationContext.cs (hunk 6 above)
Comment (a-ctor) [reply Core/MappingResolution/DefaultSetOperationReconciliationContext.cs:171]: Personal preference as I don't like the `++` operator, especially as a statement.

File: UnitTests/MappingResolution/SetOperationReconciliationVisitorTest.cs (hunk 13 above)
Comment (a-ctor) [reply UnitTests/MappingResolution/SetOperationReconciliationVisitorTest.cs:73]: The mocks are already strict and any call that is not setup will result in an exception.

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs (hunk 22 above)
Comment (a-ctor) [reply Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: We only provide the ability to create constant values so the name is fitting for me.

Comment (MichaelKetting) [reply Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: That's for now, but with the updated design we only need to add a new factory method to remove that constraint. And then we can also add computed columns using the same expression type and no breaking Change. If we stick to the current name, we need to add an entire new branch to the visitor without any benefit. Keeping the door open  makes the design more robust for this type of expansion without breaking changes. The only thing I don't have is a good name with my best guess being "named projection column expression" :) if there's a downside to this updated design , I'm all ears.

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs [hunk 24]
```diff
@@ -77,6 +77,58 @@ public void GenerateSql_VisitSqlColumnDefinitionExpression ()
...
+      Assert.That (_commandBuilder.GetCommandText(), Is.EqualTo ("@1 AS [ID]"));
+
+      var parameters = _commandBuilder.GetCommandParameters();
+      Assert.That (parameters.Length, Is.EqualTo (1));
+      Assert.That (parameters[0].Name, Is.EqualTo ("@1"));
+      Assert.That (parameters[0].Value, Is.EqualTo ("constant"));
+    }
+
+    [Test]
+    public void GenerateSql_VisitSqlConstantColumnExpression ()
```
Comment (MichaelKetting) [inline UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs]:
```suggestion
    public void GenerateSql_VisitSqlConstantColumnExpression_WithSqlLiteral ()
```

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs [hunk 25]
```diff
@@ -77,6 +77,58 @@ public void GenerateSql_VisitSqlColumnDefinitionExpression ()
...
+
+      SqlGeneratingExpressionVisitor.GenerateSql (sqlColumnExpression, _commandBuilder, _stageMock.Object);
+
+      Assert.That (_commandBuilder.GetCommandText(), Is.EqualTo ("@1 AS [ID]"));
+
+      var parameters = _commandBuilder.GetCommandParameters();
+      Assert.That (parameters.Length, Is.EqualTo (1));
+      Assert.That (parameters[0].Name, Is.EqualTo ("@1"));
+      Assert.That (parameters[0].Value, Is.EqualTo ("constant"));
+    }
```
Comment (MichaelKetting) [inline UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs:114]: I did not remember that T-SQL actually supports parameters in the projection, nor did I remember that re-linq already does parameters here. This is so great because now we can also handle strings if we want to and have a separate version sqlliteral. Things fall into place wonderfully!

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs [hunk 26]
```diff
@@ -77,6 +77,58 @@ public void GenerateSql_VisitSqlColumnDefinitionExpression ()
...
+      var sqlColumnExpression = SqlConstantColumnExpression.CreateSqlLiteral (
+          "literal",
+          typeof (int),
+          "s",
+          "ID",
+          false);
+
+      SqlGeneratingExpressionVisitor.GenerateSql (sqlColumnExpression, _commandBuilder, _stageMock.Object);
+
+      Assert.That (_commandBuilder.GetCommandText(), Is.EqualTo ("'literal' AS [ID]"));
```
Comment (MichaelKetting) [inline UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs]: Where do we get a single-quotes from here? I don't think that was your doing so there's some magic. 

Decision: we drop the CreateSqlLiteral-factory method for now because we don't need it thanks to the updated behavior of CreateConstant and I don't want to look around for the best way to solve this.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
Okay, there's just the rename of the Expression left now. plus very minor review notes on the rest.

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs (hunk 25 above)
Comment (a-ctor) [reply UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs:114]: Yes. If the columns on the entity weren't restricted to `SqlColumnExpression`, we wouldn't have needed any special class as there is already a `NamedExpression` that acts as `<expr> AS <name>`.

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs (hunk 26 above)
Comment (a-ctor) [reply UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs]: The `SqlCommandBuilder` has an `AppendStringLiteral` function that also handles basic replacement:

```cs
    public void AppendStringLiteral (string value)
    {
      Append ("'");
      Append (value.Replace ("'", "''"));
      Append ("'");
    }
```

But it is still better to drop the special case for now as there is no need as a the parameter variant covers this as well.

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs (hunk 25 above)
Comment (MichaelKetting) [reply UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs:114]: Oh, there it is. Okay. Good to know why we can't use it and nope, not gonna touch that can of worms.

File: UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs (hunk 26 above)
Comment (MichaelKetting) [reply UnitTests/SqlGeneration/SqlGeneratingExpressionVisitorTest.cs]: Drop it is.

Attention: there's a difference between "StringLiteral" and "SqlLiteralExpression". The second one is meant to inject random parts text into the SQL statement to do whatever you want, including Sql Injection. Normally, it's for supplying SQL that's not supported by re-linq natively and acts as a feature backdoor.

File: Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs (hunk 22 above)
Comment (a-ctor) [reply Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: If you want a more general name, I'd suggest derived or computed instead of constant. The use case here is not limited to the projection and the class itself could be used in other use cases as well. Derived/Computed columns match that intent quite good in my opinion.

The distinction between a computed column and a constant column might be useful (having two different classes+visitors instead of just one). If we extend in the future, there might also be a case for making a separate class. For this PR, is just need a decision on how we want to proceed.

Comment (MichaelKetting) [reply Core/SqlStatementModel/Resolved/SqlConstantColumnExpression.cs]: Decision: SqlComputedColumnExpression.

And from what I remember of relinq I'll put reputation credits on "we won't need separate types for this in the future because the polymorph value and our vistors can handle the rest"

File: UnitTests/SqlStatementModel/Resolved/SqlComputedColumnExpressionTest.cs [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+using System;
+using System.Linq.Expressions;
+using NUnit.Framework;
+using Remotion.Linq.SqlBackend.Development.UnitTesting;
+using Remotion.Linq.SqlBackend.SqlStatementModel.Resolved;
+
+namespace Remotion.Linq.SqlBackend.UnitTests.SqlStatementModel.Resolved
+{
+  [TestFixture]
+  public class SqlComputedColumnExpressionTest
```
Comment (MichaelKetting) [inline UnitTests/SqlStatementModel/Resolved/SqlComputedColumnExpressionTest.cs:27]: Please rename the file and update the commit message for the SqlContesntExpression

File: Core/SqlStatementModel/Resolved/SqlComputedColumnExpression.cs [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+using System.Linq.Expressions;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel.Resolved
+{
+  /// <summary>
+  /// Defines a SQL column as an expression with a given column name. The column is represented as "[expr] AS [name]".
+  /// The value can either be a constant or a computed value.
+  /// </summary>
+  public class SqlComputedColumnExpression : SqlColumnExpression
```
Comment (MichaelKetting) [inline Core/SqlStatementModel/Resolved/SqlComputedColumnExpression.cs:27]: We don't have TeamCity etc checks set up for Relinq SqlBackend, so please wait for Teamcity before the merge. This comment is just a tripwire to not forget this :)

**Review by MichaelKetting — APPROVED** [review]
